### ISSUE TITLE
Evaluate ML model for secret detection (DistilBERT / SAP password-model)

### ISSUE THREAD

**Issue by r000bin** [secretgate/secretgate#2]
## Summary

Evaluate fine-tuning a small transformer model for secret detection to complement the existing regex-based approach. An ML model can catch novel/unknown secret patterns that regex misses (hardcoded passwords, database connection strings, novel credential formats).

## Dataset: SecretBench

- **Source**: [Zenodo](https://doi.org/10.5281/zenodo.15273217) (CC-BY-4.0 license)
- **Paper**: [Secret Breach Detection in Source Code with LLMs](https://arxiv.org/html/2504.18784v1)
- **Size**: 97,479 candidates, 15,084 confirmed true secrets
- **Categories**: Private Keys (5,791), API Keys (4,521), Auth Tokens (3,567), Generic Secrets (333), DB/Server URLs (162), Passwords (150), Usernames (27), Other (524)
- **Coverage**: 49 programming languages, 311 file types

## Model Candidates

| Model | Params | Size (quantized) | CPU Latency (ONNX) | Code-aware |
|-------|--------|-------------------|---------------------|------------|
| BERT-Tiny | ~4.4M | <50 MB | 10-20ms | No |
| DistilBERT | 66M | <100 MB | 30-50ms | No |
| CodeBERT | 125M | ~120 MB | 100-200ms | Yes |
| **SAP/password-model** | 125M | ~120 MB | 100-200ms | Yes (pre-fine-tuned) |

### Existing model: SAP/password-model
- Already fine-tuned on CodeBERT for credential/leak detection
- Used in SAP's [Credential Digger](https://github.com/SAP/credential-digger) tool
- Available on HuggingFace: https://huggingface.co/SAP/password-model
- **Should be evaluated first before training from scratch**

## Integration Architecture

```
Request
  |
  v
[Regex scan] -----> known patterns (fast, <1ms)
  |
  v
[ML model] -------> novel/unknown patterns (30-50ms with DistilBERT+ONNX)
  |
  v
[Decision: block/audit/pass]
  |
  v
Forward to LLM
```

## Expected Performance

The paper achieved F1=0.9852 with fine-tuned LLaMA-3.1 8B. A DistilBERT-class model should achieve F1 ~0.93-0.96 with <50ms latency on CPU (ONNX INT8 quantized). This latency is negligible compared to LLM response times.

## Resource Requirements

| Resource | DistilBERT (ONNX INT8) | CodeBERT/SAP model |
|----------|------------------------|---------------------|
| RAM | ~200 MB | ~500 MB |
| Disk | ~100 MB | ~250 MB |
| CPU cores | 1-2 | 2-4 |
| Latency/request | 30-50ms | 100-200ms |

## Risks

- **Class imbalance**: Passwords (150) and Usernames (27) are severely underrepresented — keep regex for those categories
- **False positives**: Start in audit mode, tune threshold before enforcing
- SecretBench data may require running extraction scripts to obtain full dataset

## Suggested Plan

1. Download SecretBench, evaluate SAP/password-model against it
2. If insufficient, fine-tune DistilBERT or CodeBERT on SecretBench (binary classification: secret vs not-secret)
3. ONNX export + INT8 quantization
4. Integrate into secretgate as optional ML detection layer
5. Validate against SecretBench test set + manual testing with known secret patterns
